File: Server/go/apptoken/apptokenOptions.go
package apptoken

import (
	"bytes"
	"encoding/binary"
	"errors"
	"io"
	"sort"
)

type AppTokenOptions struct {
	EngineOptions map[string]string
}

func CreateAppTokenOptions() AppTokenOptions {
	return AppTokenOptions{
		EngineOptions: make(map[string]string),
	}
}

func (options *AppTokenOptions) SetEngineOptions(engineOptions map[string]string) {
	options.EngineOptions = engineOptions
}

func (options *AppTokenOptions) Pack() ([]byte, error) {
	buf := new(bytes.Buffer)

	// hasEngineOptions
	hasEngineOptions := options.EngineOptions != nil
	if err := binary.Write(buf, binary.BigEndian, hasEngineOptions); err != nil {
		return nil, err
	}
	if hasEngineOptions {
		if err := binary.Write(buf, binary.BigEndian, int32(len(options.EngineOptions))); err != nil {
			return nil, err
		}

		if len(options.EngineOptions) > 0 {
			// sort by key
			keys := make([]string, 0, len(options.EngineOptions))
			for k := range options.EngineOptions {
				keys = append(keys, k)
			}
			sort.Strings(keys)

			for _, key := range keys {
				value := options.EngineOptions[key]
				if key == "" || value == "" {
					return nil, errors.New("illegal engineOptions entry")
				}
				if err := binary.Write(buf, binary.BigEndian, int32(len(key))); err != nil {
					return nil, err
				}
				if _, err := buf.Write([]byte(key)); err != nil {
					return nil, err
				}

				if err := binary.Write(buf, binary.BigEndian, int32(len(value))); err != nil {
					return nil, err
				}
				if _, err := buf.Write([]byte(value)); err != nil {
					return nil, err
				}
			}
		}
	}

	return buf.Bytes(), nil
}

func UnpackAppTokenOptions(buf io.Reader) (*AppTokenOptions, error) {
	options := &AppTokenOptions{
		EngineOptions: make(map[string]string),
	}
	// hasEngineOptions
	var hasEngineOptions bool
	if err := binary.Read(buf, binary.BigEndian, &hasEngineOptions); err != nil {
		return nil, err
	}

	if hasEngineOptions {

		var size int32
		if err := binary.Read(buf, binary.BigEndian, &size); err != nil {
			return nil, err
		}

		for i := int32(0); i < size; i++ {
			var keyLength int32
			if err := binary.Read(buf, binary.BigEndian, &keyLength); err != nil {
				return nil, err
			}
			key := make([]byte, keyLength)
			if _, err := buf.Read(key); err != nil {
				return nil, err
			}

			var valueLength int32
			if err := binary.Read(buf, binary.BigEndian, &valueLength); err != nil {
				return nil, err
			}
			value := make([]byte, valueLength)
			if _, err := buf.Read(value); err != nil {
				return nil, err
			}

			options.EngineOptions[string(key)] = string(value)
		}
	}

	return options, nil
}

File: Server/csharp/AppToken/Model/AppTokenOptions.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using AppToken.Util;

namespace AppToken.Model
{
    public class AppTokenOptions
    {
        private Dictionary<string, string> engineOptions;

        public AppTokenOptions()
        {
            engineOptions = new Dictionary<string, string>();
        }

        private AppTokenOptions(Dictionary<string, string> engineOptions)
        {
            this.engineOptions = engineOptions ?? throw new ArgumentNullException(nameof(engineOptions));
        }

        public void SetEngineOptions(Dictionary<string, string> engineOptions)
        {
            this.engineOptions = engineOptions ?? throw new ArgumentNullException(nameof(engineOptions));
        }

        public void Pack(MemoryStream buf)
        {
            using (var writer = new BinaryWriter(buf, Encoding.UTF8, true))
            {
                writer.Write(engineOptions != null); // Write bool for existence
                if (engineOptions == null)
                {
                    return;
                }

                writer.Write(BigEndianUtils.GetBytesBigEndian(engineOptions.Count));
                var sortedKeys = engineOptions.Keys.OrderBy(k => k).ToList();
                foreach (var key in sortedKeys)
                {
                    var value = engineOptions[key];
                    if (key == null || value == null)
                        throw new ArgumentException("Illegal engineOptions entry");

                    writer.Write(BigEndianUtils.GetBytesBigEndian(key.Length));
                    writer.Write(Encoding.UTF8.GetBytes(key));

                    writer.Write(BigEndianUtils.GetBytesBigEndian(value.Length));
                    writer.Write(Encoding.UTF8.GetBytes(value));
                }
            }
        }

        public static AppTokenOptions Unpack(MemoryStream buf)
        {
            var engineOptions = new Dictionary<string, string>();

            using (var reader = new BinaryReader(buf, Encoding.UTF8, true))
            {
                if (!reader.ReadBoolean())
                    return new AppTokenOptions();

                int size = BigEndianUtils.FromBytesBigEndianToInt(reader.ReadBytes(4));
                for (int i = 0; i < size; i++)
                {
                    int keyLength = BigEndianUtils.FromBytesBigEndianToInt(reader.ReadBytes(4));
                    string key = Encoding.UTF8.GetString(reader.ReadBytes(keyLength));

                    int valueLength = BigEndianUtils.FromBytesBigEndianToInt(reader.ReadBytes(4));
                    string value = Encoding.UTF8.GetString(reader.ReadBytes(valueLength));

                    engineOptions.Add(key, value);
                }
            }

            return new AppTokenOptions(engineOptions);
        }


    }
}